Add C++ Major Mode

mayanez opened the pull request:
Adds C++ Major Mode Key Bindings. (See https://github.com/VSpaceCode/VSpaceCode/issues/98)

I've tried to follow the Spacemacs keybindings as much as possible. I will be updating this as I figure out what other key bindings are useful to add. There's still a few missing things missing, so if anyone has more suggestions as what to add please comment.

stevenguh commented:
Looking at https://github.com/microsoft/vscode-cpptools/blob/main/Extension/package.json#L1163-L1267 for all the available commands.

Maybe `C_Cpp.ResetDatabase` can be bound to `SPC m b f` as [refresh index](https://develop.spacemacs.org/layers/+lang/c-c++/README.html#lsp-1)

Two follow ups:
1. Can you send another PR to add cpp in https://github.com/VSpaceCode/vspacecode.github.io/blob/master/docs/major-mode.md?
2. Can you run `npm run sort-bindings` to sort the bindings?

mayanez commented:
Sounds good. I will tackle this on the weekend.

mayanez commented:
> Maybe `C_Cpp.ResetDatabase` can be bound to `SPC m b f` as [refresh index](https://develop.spacemacs.org/layers/+lang/c-c++/README.html#lsp-1)

I added an `IntelliSense` category and added the corresponding `ResetDatabase` and `RescanWorkspace` commands.

stevenguh commented on package.json:
```diff
@@ -2821,6 +2821,131 @@
+												]
+											},
+											{
+												"key": "i",
+												"name": "+IntelliSense",
```
```suggestion
												"key": "b",
												"name": "+Backend",
```
https://develop.spacemacs.org/layers/+tools/lsp/README.html
Should we use `b` and "Backend" for this to be consistent? We use `b` for this purpose in other language's major mode as well.

mayanez replied:
Sure. If helps keep consistency across all layers I'd be happy to update it. Which two bindings should I updated it with.

mayanez replied:
Updated to use `b` for consistency.

stevenguh commented on package.json:
```diff
@@ -2821,6 +2821,131 @@
+												"type": "bindings",
+												"bindings": [
+													{
+														"key": "=",
+														"name": "Format Document",
+														"type": "command",
+														"command": "editor.action.formatDocument"
+													},
+													{
+														"key": "s",
+														"name": "Format Selection",
+														"type": "command",
+														"command": "editor.action.formatSelection"
+													}
```
Is there any reason these are not following https://vspacecode.github.io/docs/conventions?
Note: not all LSP provide all format options. Make sure you test it and make sure c++ works with the options.

mayanez replied:
I was oblivious to this existing thanks for calling it out. I've tested and updated them to match the conventions.

stevenguh commented:
I just merged my fix of the sorting script #179. Please merge to master, and run the sorting script again. Thanks :)

Note: Merging is probably easier than rebasing, and I usually do squash and merge for PR anyway.

stevenguh commented on package.json:
```diff
@@ -1813,6 +1813,155 @@
+												"name": "+Refactor",
+												"type": "bindings",
+												"bindings": [
+													{
+														"key": "s",
```
```suggestion
													{
														"key": "r",
```
Other language uses `r`. Probably good to follow?

mayanez replied:
Sure. I'll make this change.

stevenguh commented on package.json:
```diff
@@ -1813,6 +1813,155 @@
+														"command": "C_Cpp.SwitchHeaderSource"
+													},
+													{
+														"key": "d",
+														"name": "Peek declaration",
```
Is there any reason it seems like these bindings are inverse of other bindings (e.g. go's bindings)
d is usually go to definition instead of peek
g is usually go to definition instead of peek

stevenguh replied:
Got you. I thought vscode usually do a peek if there's more than one references/definitions/declarations when you do a reveal. If that's not the case, I think what you are proposing makes sense.

@MarcoIeni Any thoughts on this?

stevenguh replied:
In my limited session with C++, it does "intelligently" show the peek overlay.

@mayanez How about we flip the peek and reveal you have here, so it's more consistent across languages. If we have more complaints, we can always change it 😄

MarcoIeni replied:
Yeah, I find peek annoying, what about moving current bindings to upper case `D` and `G`  and having `d` and `g` to `editor.action.revealDefinition` as golang?

mayanez replied:
Sure I will flip them then. Not a big deal.

stevenguh reviewed (commented):
Sorry for another round of comments. That should be the last.

stevenguh commented on package.json:
```diff
@@ -1813,6 +1813,155 @@
+														"key": "r",
+														"name": "Peek references",
+														"type": "command",
+														"command": "editor.action.referencesSearch.trigger"
```
```suggestion
														"command": "editor.action.referenceSearch.trigger"
```

mayanez commented:
Rebased and updated the peek/reveal bindings.
